### PR TITLE
Load gradle properties earlier

### PR DESCRIPTION
**Pull request by eskatos**
As early as required for both the configuration cache and file watching on the root build while keeping the system properties mutation at the same point in time, and without changing nested builds behaviour.

This allows to check for system properties persistently set in `gradle.properties` as `systemProp.something` very early. They aren't available as `System.getProperty("something")` yet but can be queried on `GradleProperties.find("systemProp.something")`.

Without loading the properties earlier the behaviour is quite undefined. Observations are that if you are using [the wrapper](https://github.com/gradle/gradle/blob/982d3999a0fbb9aa1a236a517e50c3e95631f925/subprojects/wrapper/src/main/java/org/gradle/wrapper/GradleWrapperMain.java#L69-L72) you can query early for system properties defined in `gradle.properties` files but not when not using the wrapper or using TAPI, from an IDE for example.

Proper build options (`--configuration-cache` or `--watch-fs`) don't require this change.

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/initialization/DefaultGradlePropertiesLoader.java [hunk 1]
```diff
@@ -79,6 +76,20 @@ private void addGradleProperties(Map<String, String> target, File... files) {
...
     }
 
+    private Map<String, String> getPropertiesFromSystemPropertiesArgs(StartParameterInternal startParameter) {
+        Map<String, String> properties = new HashMap<>();
+        for (Map.Entry<String, String> entry : startParameter.getSystemPropertiesArgs().entrySet()) {
+            String key = entry.getKey();
+            if (key.startsWith(SYSTEM_PROJECT_PROPERTIES_PREFIX) && key.length() > SYSTEM_PROJECT_PROPERTIES_PREFIX.length()) {
+                properties.put(key.substring(SYSTEM_PROJECT_PROPERTIES_PREFIX.length()), entry.getValue());
+            } else {
+                properties.put(Project.SYSTEM_PROP_PREFIX + '.' + key, entry.getValue());
```
[the remark this answers is not in the document]
Comment (eskatos) [reply subprojects/core/src/main/java/org/gradle/initialization/DefaultGradlePropertiesLoader.java]: Done

File: subprojects/composite-builds/src/main/java/org/gradle/composite/internal/DefaultRootBuildState.java [hunk 2]
```diff
@@ -107,4 +112,18 @@ public Path getCurrentPrefixForProjectsInChildBuilds() {
     public Path getIdentityPathForProject(Path path) {
         return path;
     }
+
+    private void loadGradleProperties() {
```
[the remark this answers is not in the document]
Comment (eskatos) [reply subprojects/composite-builds/src/main/java/org/gradle/composite/internal/DefaultRootBuildState.java:116]: As discovered in the last attempts, see my comments in #13075, using an injected `BuildLayout`  has some side effects changing the behaviour for composite builds by letting root build properties shadow included builds properties. Loading in the constructor is also problematic as it has some side effects changing the behaviour for project properties in nested builds such as `buildSrc`.

[the remark this answers is not in the document]
Comment (adammurdoch) [reply subprojects/composite-builds/src/main/java/org/gradle/composite/internal/DefaultRootBuildState.java:116]: I'm not sure what is a "build property" is and what shadowing you mean. There are system properties and there are Gradle properties. Each build has its own controller and its own layout, and the controller loads the Gradle and system properties based on this. At some point in time, the controller mutates the contents of `System.getProperties()`, but this point in time doesn't change by using the injected layout.

[the remark this answers is not in the document]
Comment (eskatos) [reply subprojects/composite-builds/src/main/java/org/gradle/composite/internal/DefaultRootBuildState.java:116]: I meant "root build - project properties" vs "included builds - project properties". See failing tests linked in https://github.com/gradle/gradle/pull/13075#issuecomment-628556476, in particular https://e.grdev.net/s/xdip6ag2a3ur6/tests/failed

**Comment by eskatos** [comment]
Superseded by https://github.com/gradle/gradle/pull/16757

**Pull request closed without merging**